### PR TITLE
New identification rule for entity columns

### PR DESCRIPTION
**Pull request by dmytro-grankin**
This PR brings the new rule for identification of entity columns. Since, besides the old rules, an entity column should be annotated with `javax.persistence.Column`.

### CODE REVIEW

**Comment by dmytro-grankin** [comment]
@dmdashenkov PTAL.

File: server/src/main/java/io/spine/server/entity/AbstractEntity.java [hunk 1]
```diff
@@ -85,12 +85,11 @@ public I getId() {
...
     @CheckReturnValue
     public S getState() {
-        S result = state;
-        if (result == null) {
+        if (state == null) {
             state = getDefaultState();
-            result = state;
         }
-        return result;
+
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/AbstractEntity.java]: Please remove the empty line.

File: server/src/main/java/io/spine/server/entity/storage/Column.java [hunk 2]
```diff
@@ -76,16 +79,23 @@
  *      {@code
  *         --UserAggregate.java--
  *
+ *         // only methods annotated with @javax.persistence.Column may be considered Columns
+ *         public Department getDepartment() { ... }
+ *
  *         // non-public methods may not represent Columns
+ *         \@Column
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Column.java]: Please check if this form is interpreted correctly.

File: server/src/main/java/io/spine/server/entity/storage/Column.java [hunk 3]
```diff
@@ -108,8 +118,7 @@
  * <h2>Nullability</h2>
  *
  * <p>A Column may turn into {@code null} value if the getter which declares it is annotated as
- * {@link javax.annotation.Nullable javax.annotation.Nullable}. Otherwise, the Column is considered
- * non-null.
+ * {@link javax.annotation.Nullable}. Otherwise, the Column is considered non-null.
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Column.java]: Please revert this change. Spine Javadoc style checker would consider this as an error.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 4]
```diff
@@ -47,7 +50,12 @@
  *
  * <p>All the methods of the passed {@link Entity} that fit
  * <a href="http://download.oracle.com/otndocs/jcp/7224-javabeans-1.01-fr-spec-oth-JSpec/">
- * the Java Bean</a> getter spec are considered {@link Column Columns}.
+ * the Java Bean</a> getter spec and annotated with {@link javax.persistence.Column}
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: Same as above: use `{@link f.q.N f.q.N}` form.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 5]
```diff
@@ -214,14 +207,76 @@ private static void addToCache(Class<? extends Entity> entityType) {
...
+     *
+     * @param getter the getter to check
+     * @return {@code true} if the method is getter for {@link Column}
+     */
+    private static boolean isEntityColumn(Method getter) {
+        final boolean isAnnotated = getter.isAnnotationPresent(javax.persistence.Column.class);
+        if (isAnnotated) {
+            return true;
+        }
+
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: Please remove the empty line.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 6]
```diff
@@ -214,14 +207,76 @@ private static void addToCache(Class<? extends Entity> entityType) {
...
+        for (Class<?> ascendant : ascendants) {
+            final Optional<Method> optional = getMethod(getter, ascendant);
+            if (optional.isPresent()) {
+                final Method ascendantMethod = optional.get();
+                if (ascendantMethod.isAnnotationPresent(javax.persistence.Column.class)) {
+                    return true;
+                }
+            }
+        }
+
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: Same as above.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 7]
```diff
@@ -214,14 +207,76 @@ private static void addToCache(Class<? extends Entity> entityType) {
...
+    }
+
+    /**
+     * Obtains the method with the same signature as the specified method from the specified class.
+     *
+     * @param method the method to get the signature
+     * @param aClass the class to obtain the method with the same signature
+     * @return the method with the same signature obtained from the specified class
+     */
+    private static Optional<Method> getMethod(Method method, Class<?> aClass) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: Please find a better name for `aClass`.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 8]
```diff
@@ -214,14 +207,76 @@ private static void addToCache(Class<? extends Entity> entityType) {
...
+                if (ascendantMethod.isAnnotationPresent(javax.persistence.Column.class)) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private static Iterable<Class<?>> getSuperClassesAndInterfaces(Class<?> aClass) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: Please find a better name for `aClass`.

Comment (dmdashenkov) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: `cls` would fit in this case.

File: server/src/main/java/io/spine/server/entity/storage/Column.java (hunk 3 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Column.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 4 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 5 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 6 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 8 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 7 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Column.java (hunk 2 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Column.java]: Javadocs will contain backslashes, but there are no better options due to the [bug](https://bugs.openjdk.java.net/browse/JDK-8130754).

**Comment by dmytro-grankin** [comment]
@dmdashenkov PTAL.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 9]
```diff
@@ -50,8 +50,8 @@
  *
  * <p>All the methods of the passed {@link Entity} that fit
  * <a href="http://download.oracle.com/otndocs/jcp/7224-javabeans-1.01-fr-spec-oth-JSpec/">
- * the Java Bean</a> getter spec and annotated with {@link javax.persistence.Column}
- * are considered {@link Column Columns}.
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: `{@linkplain }`

File: server/src/main/java/io/spine/server/entity/storage/Columns.java [hunk 10]
```diff
@@ -258,19 +256,19 @@ private static boolean isEntityColumn(Method getter) {
...
-     * Obtains the method with the same signature as the specified method from the specified class.
+     * Obtains the method from the specified class with the same signature as the specified method.
      *
+     * @param target the class to obtain the method with the signature
      * @param method the method to get the signature
-     * @param aClass the class to obtain the method with the same signature
      * @return the method with the same signature obtained from the specified class
      */
-    private static Optional<Method> getMethod(Method method, Class<?> aClass) {
+    private static Optional<Method> getMethod(Class<?> target, Method method) {
```
Comment (dmdashenkov) [inline server/src/main/java/io/spine/server/entity/storage/Columns.java]: Can the result method be non-`public`? 
Please check and add a corresponding test.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 9 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

File: server/src/main/java/io/spine/server/entity/storage/Columns.java (hunk 10 above)
Comment (dmytro-grankin) [reply server/src/main/java/io/spine/server/entity/storage/Columns.java]: Done.

**Comment by dmytro-grankin** [comment]
@dmdashenkov PTAL.
